### PR TITLE
add-relative-functions-to-signals -> master

### PR DESCRIPTION
**Pull request by Nitay4**

### CODE REVIEW

File: src/main/java/frc/robot/hardware/signal/AngleSignal.java [hunk 1]
```diff
@@ -43,6 +44,26 @@ public double[] getTimestamps() {
...
 	}
 
+	@Override
+	public boolean isNear(Rotation2d value, Rotation2d tolerance) {
+		return MathUtil.isNear(value.getRotations(), getLatestValue().getRotations(), tolerance.getRotations());
+	}
+
+	@Override
+	public boolean isFarBy(Rotation2d value, Rotation2d distance) {
+		return !MathUtil.isNear(value.getRotations(), getLatestValue().getRotations(), distance.getRotations());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/AngleSignal.java:54]: use the function isNear from above

File: src/main/java/frc/robot/hardware/signal/DoubleSignal.java [hunk 2]
```diff
@@ -39,6 +40,26 @@ public double[] getTimestamps() {
...
 	}
 
+	@Override
+	public boolean isNear(Double value, Double tolerance) {
+		return MathUtil.isNear(value, getLatestValue(), tolerance);
+	}
+
+	@Override
+	public boolean isFarBy(Double value, Double distance) {
+		return !MathUtil.isNear(value, getLatestValue(), distance);
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/DoubleSignal.java:50]: use the function is near from above

File: src/main/java/frc/robot/hardware/signal/DoubleSignal.java [hunk 3]
```diff
@@ -39,6 +40,26 @@ public double[] getTimestamps() {
...
+	}
+
+	@Override
+	public boolean isMore(Double value) {
+		return getLatestValue() > value;
+	}
+
+	@Override
+	public boolean isLess(Double value) {
+		return getLatestValue() > value;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/signal/DoubleSignal.java:60]: switch the sign

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/signal/AngleSignal.java [hunk 4]
```diff
@@ -43,6 +44,26 @@ public double[] getTimestamps() {
...
 
+	@Override
+	public boolean isNear(Rotation2d value, Rotation2d tolerance) {
+		return MathUtil.isNear(value.getRotations(), getLatestValue().getRotations(), tolerance.getRotations());
+	}
+
+	@Override
+	public boolean isFarBy(Rotation2d distance) {
+		return !isNear(distance, Rotation2d.fromRotations(0));
+	}
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/signal/AngleSignal.java:55]: this is not how math works.
also u want to get value and return if the distnace of current to given value is bigger that given distnace

**Review by YHerm — CHANGES_REQUESTED** [review]
Please test it in simulation

File: src/main/java/frc/robot/hardware/interfaces/InputSignal.java [hunk 5]
```diff
@@ -14,4 +14,12 @@ public interface InputSignal<T> extends LoggableInputs {
 
 	double[] getTimestamps();
 
+	boolean isNear(T value, T tolerance);
+
+	boolean isFarBy(T value, T distance);
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/interfaces/InputSignal.java:19]: why `By`? you don't do the same for `isNear`

File: src/main/java/frc/robot/hardware/interfaces/InputSignal.java [hunk 6]
```diff
@@ -14,4 +14,12 @@ public interface InputSignal<T> extends LoggableInputs {
...
+
+	boolean isMore(T value);
+
+	boolean isLess(T value);
```
Comment (Yotam7) [inline src/main/java/frc/robot/hardware/interfaces/InputSignal.java:23]:
```suggestion
	boolean isLessThan(T value);
```

File: src/main/java/frc/robot/hardware/signal/DoubleSignal.java [hunk 7]
```diff
@@ -39,6 +40,26 @@ public double[] getTimestamps() {
 		return new double[] {timestamp};
 	}
 
+	@Override
+	public boolean isNear(Double value, Double tolerance) {
+		return MathUtil.isNear(value, getLatestValue(), tolerance);
+	}
+
+	@Override
+	public boolean isFar(Double value, Double distance) {
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/signal/DoubleSignal.java:49]: isFurthur

File: src/main/java/frc/robot/hardware/signal/AngleSignal.java [hunk 8]
```diff
@@ -43,6 +44,27 @@ public double[] getTimestamps() {
...
 	}
 
+	@Override
+	public boolean isNear(Rotation2d value, Rotation2d tolerance) {
+		// double expected, double actual, double tolerance
+		return MathUtil.isNear(value.getRotations(), getLatestValue().getRotations(), tolerance.getRotations());
+	}
+
+	@Override
+	public boolean isFar(Rotation2d value, Rotation2d distance) {
```
Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/signal/AngleSignal.java:54]: isFurthur

Comment (NoamRosenberg08) [inline src/main/java/frc/robot/hardware/signal/AngleSignal.java:54]: isFurthur

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/interfaces/InputSignal.java [hunk 9]
```diff
@@ -14,4 +14,12 @@ public interface InputSignal<T> extends LoggableInputs {
 
 	double[] getTimestamps();
 
+	boolean isNear(T value, T tolerance);
+
+	boolean isFurthur(T value, T distance);
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/interfaces/InputSignal.java:19]: furthur? do you mean further?

**Review by NoamRosenberg08 — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/signal/AngleSignal.java [hunk 10]
```diff
@@ -43,6 +44,26 @@ public double[] getTimestamps() {
 		return new double[] {timestamp};
 	}
 
+	@Override
+	public boolean isNear(Rotation2d value, Rotation2d tolerance) {
+		return MathUtil.isNear(value.getRotations(), getLatestValue().getRotations(), tolerance.getRotations());
+	}
+
+	@Override
+	public boolean isFurthur(Rotation2d value, Rotation2d distance) {
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/signal/AngleSignal.java:53]:
```suggestion
	public boolean isFurthur(Rotation2d value, Rotation2d tolerance) {
```

File: src/main/java/frc/robot/hardware/signal/DoubleSignal.java [hunk 11]
```diff
@@ -39,6 +40,26 @@ public double[] getTimestamps() {
 		return new double[] {timestamp};
 	}
 
+	@Override
+	public boolean isNear(Double value, Double tolerance) {
+		return MathUtil.isNear(value, getLatestValue(), tolerance);
+	}
+
+	@Override
+	public boolean isFurthur(Double value, Double distance) {
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/signal/DoubleSignal.java:49]:
```suggestion
	public boolean isFurthur(Double value, Double tolerance) {
```

**Review by hillelv — CHANGES_REQUESTED** [review]

**Review by hillelv — APPROVED** [review]

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by DANNA2828 — APPROVED** [review]
(/*0*)/

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by tnrz135 — APPROVED** [review]

**Review by YHerm — APPROVED** [review]

**Pull request merged**